Add pyright and cleanup the codebase

ItsDrike opened the pull request:
This adds pyright type checker to enforce proper type-hinting and type propagation everywhere. With it, the `StrictABC`'s type-enforcement functionalities are no longer necessary, and so they will be removed.

I've also done some general codeabse cleanup for non-ideal things I found in the code while making the project pyright compliant and fixing several type-hints. These cleanup changes sometimes even fix major bugs, which went previously unnoticed but were catched by pyright. Note that these bugs would never make it into production if the test coverage was better.

Tasks:
- [x] Add [python pyright](https://pypi.org/project/pyright/) development dependency to avoid the need for manual NPM installation.
- [ ] Add mention of pyright enforcement and how to use it into `CONTRIBUTING.md`
- [ ] **(WIP)** Make the codebase pyright compliant
- [ ] **(WIP)** Do some additional cleanup work
- [ ] Add automatic github workflow to ensure pyright is passing

ItsDrike committed:
Add pyright dev dependency

ItsDrike committed:
Make registry typed
This also fixes a slight issue where if a list/tuple was passed into the
registry class as data, it was treated as the reversed data rather, and
reversed data was the actual data.

ItsDrike committed:
Improve from_list logic and fix type hints
For some reason, the from_list method was originally making the instance
in which it then (optionally) checked for duplicate packets based on ID.
However during the original parsing, packets were stored as dicts with
IDs as keys, this means even if there were some duplicates, they'd just
get overridden and we'd only pass the last packet in the list with that
ID. This meant the duplicate check would never actually catch any
duplicate packets.

This logic was also not ideal because it went through the list of
packets server times for no reason, when it could all be done in a
single loop. This is way more efficient which would become noticable if
the list of packets was long enough.

It also fixes an invalid type hint, which specified the packets argument
should be a List[Type[Packet]], however this would mean we expect the
list to contain the actual "Packet" class (or it's subtypes), when in
fact we want instances of this type. Not only that, but the function
can't even process any other type of packets than ClientBound and
ServerBound ones, so we shouldn't specifiy that it takes any arbitrary
packet, but rather a union of the two.

ItsDrike committed:
Enum comparison should be is, not __eq__

ItsDrike committed:
Take __getitem__ argument as positional-only

ItsDrike committed:
Use a more specific return type

ItsDrike committed:
Specify __getitems__ overload

ItsDrike committed:
Fix overload type hint

ItsDrike committed:
Fix from_list fucntion
In 7dcbfa9, I wrongly assumed that the function was taking instances,
when in fact it was taking classes, this reverts that part of the
commit.

ItsDrike committed:
Fix return type for __getitem__
Originally, this return type was "Packet, which I've then change to a
Union of 2 subtypes of Packet, however this type was actually supposed
to be a Type[Packet], so this changes the return type to the type
(class) of the union of these 2 subtypes

ItsDrike committed:
Improve error message

ItsDrike committed:
Follow black

456dev commented:
I mean pyright requires node js etc, and that python package would install it as well
It's not my favourite idea.

Have the python native options, like mypy (iirc) been considered?

ItsDrike commented:
> I mean pyright requires node js etc, and that python package would install it as well
It's not my favourite idea.

i mean, npm will only be installed in the python's virtual environment (assuming you did use the python pyright package for installation with poetry), not on the machine's global environment, so you shouldn't have any issues with it colliding with your system versions, or with it cluttering your system or anything like that.

> Have the python native options, like mypy (iirc) been considered?

I do know about other alternatives, written directly in python, such as mypy or even less common things like pytype or pyre. However most of these alternatives are a lot slower than pyright, and have way worse support when it comes to newer typing features (such as `typing_extensions.Self`, `typing_extensions.TypeVarTuple`, ...). Not only that, but they're often a lot harder to please and they require some very strict rules which we may not necessarily want to have to follow. Or for example in pytype's I found it's rules to not be strict enough and using it often results in some bad typing practices and some errors go unaddressed.

@the456gamer

ItsDrike committed:
Actually return Buffer from write_ingredient

ItsDrike committed:
Cast some values where necessary

ItsDrike committed:
Actually return Buffer from write_slot

ItsDrike committed:
Use Self as return type

ItsDrike committed:
Use Optional for params defaulting to None

ItsDrike committed:
Properly define __repr__

ItsDrike committed:
Fix typo

ItsDrike committed:
Properly annotate read_optional

ItsDrike committed:
Use a dict switch instead of elif mess

ItsDrike committed:
Add not-implemented write_block for refs

ItsDrike committed:
Improve docstring

ItsDrike committed:
Cast read('f') output as float

ItsDrike committed:
Run black

ItsDrike committed:
Use Optional for params defaulting to None

ItsDrike committed:
Make a more specific type for json-like values

ItsDrike committed:
Add some TODOs to fix weird things

ItsDrike committed:
Merge branch 'main' into add-pyright

ItsDrike committed:
Merge branch 'main' into add-pyright

ItsDrike committed:
Revert typing generic additions to Registry
Turns out it's actually very hard, and perhaps impossible to properly
type-hint the Registry class because python's typing doesn't have
support for Higher Kinded TypeVars. The class is also very complex (even
if it doesn't have that much code) and allows for way too many edge
cases, which would probably result in a bunch of overloads and type
ignores, making the code even more convoluted than it already is.

Instead, this removes all of the generic behavior to the Registry class
and improves some minor code quality issues. This isn't ideal, since it
means it'd going to rely on typing.Any, but due to this difficulty and
complexity, it's justified.

ItsDrike committed:
Split logic for taking sequences to a classmethod
Accepting both a mapping and a sequence from init as data is weird,
it just makes the code more messy and it's not a necessary thing that
needs to be supported by the main constructor, it's an alternative way
to pass data over, which means it should use an alternative constructor.

This adds `from_sequence` classmethod constructor to address this.

ItsDrike committed:
Properly annotate list of tag classes

ItsDrike committed:
Fix get function annotation

ItsDrike committed:
Don't make packet id optional

ItsDrike committed:
Adhere to pyright's strictParameterNoneValue
Since pyright 1.1.224, strictParameterNoneValue is now true by default,
causing typing errors on things like `x: int = None` which instead
requires explicit: `x: Optional[int] = None` to pass now.

ItsDrike committed:
Run isort

ItsDrike committed:
Allow tag names to be None

ItsDrike committed:
Cast types from Buffer.read

ItsDrike committed:
Actually return buffer from PlayPlayerInfo.pack

ItsDrike committed:
Actually initialize buffer before calling instance methods

ItsDrike committed:
Actually pass kwargs to fucntion that takes kwargs

ItsDrike committed:
Use correct type (ServerBoundPacket, not ClientBoundPacket)

ItsDrike committed:
Type-ignore vars with async init set to None in constructor

ItsDrike committed:
Don't use len() on int variable
